File: 16.3sum-closest/solution.cpp
class Solution {
public:
	int threeSumClosest(vector<int>& nums, int target) {
		sort(nums.begin(), nums.end());
		int len = nums.size();
		int min = target > 0 ? INT_MAX : INT_MIN + 1;
		for (int i = 0; i < len; i++)
		{
			if (nums[i] > (target + 2) / 3)
			{
				if (i < len - 2)
					min = nearest(min, nums[i] + nums[i + 1] + nums[i + 2], target);
				break;
			}
			if(i > 0 && nums[i] == nums[i - 1])
				continue;
			int j = i + 1;
			int k = len - 1;
			while (j < k)
			{
				int result = nums[i] + nums[j] + nums[k];
				if (result == target)
					return target;
				min = nearest(min, nums[i] + nums[j] + nums[k], target);
				if (result > target)
					k--;
				else
					j++;

			}
		}
		return min;
	}

	inline int nearest(const int &a, const int &b, const int &target)
	{
		return abs(a - target) > abs(b - target) ? b : a;
	}
};

File: 16.3sum-closest/solution.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        nearest = 100000
        for i in range(len(nums)):
            target2 = target - nums[i]
            j = 0
            k = len(nums) - 1
            if j == i:
                j += 1
            if k == i:
                k -= 1
            nearest2 = 100000
            while j < k:
                sum2 = nums[j] + nums[k]
                if sum2 == target2:
                    return target
                if abs(sum2 - target2) < abs(nearest2 - target2):
                    nearest2 = sum2
                if sum2 < target2:
                    j += 1
                    if j == i:
                        j += 1
                else:
                    k -= 1
                    if k == i:
                        k -= 1
            if abs(nearest2 + nums[i] - target) < abs(nearest - target):
                nearest = nearest2 + nums[i]
        return nearest
